### ISSUE TITLE
Error Building

### ISSUE THREAD

**Issue by ccravens** [warmcat/libwebsockets#1077]
I get the following error when attempting to build libwebsockets from source with the following commands:

```bash
echo "Building libwebsockets..."
cd libwebsockets
mkdir build
cd build
cmake .. -DCMAKE_INSTALL_PREFIX:PATH=libwebsockets/ -DLWS_WITH_PLUGINS=On -DLWS_WITH_LIBUV=On -DLWS_LIBUV_INCLUDE_DIRS:PATH=$DEPS_LIB/$LIBUV_DIR/include -DLWS_LIBUV_LIBRARIES:PATH=$DEPS_LIB/$LIBUV_DIR/lib/libuv.$LIB_EXT -DOPENSSL_ROOT_DIR=$DEPS_LIB/$OPENSSL_DIR -DCMAKE_INCLUDE_DIRECTORIES_PROJECT_BEFORE=$DEPS_LIB/$OPENSSL_DIR -DOPENSSL_EXECUTABLE=$DEPS_LIB/$OPENSSL_DIR/bin/openssl -DOPENSSL_LIBRARIES:PATH=$DEPS_LIB/$OPENSSL_DIR/lib -DZLIB_INCLUDE_DIR:PATH=$DEPS_LIB/$ZLIB_DIR/include -DZLIB_LIBRARY:FILEPATH=$DEPS_LIB/$ZLIB_DIR/lib/libz.dylib -DLWS_WITH_HTTP2=1 2>> $DEPS_LOG/$LIBWEBSOCKETS_DIR.error 1>> $DEPS_LOG/$LIBWEBSOCKETS_DIR.log
make 2>> $DEPS_LOG/$LIBWEBSOCKETS_DIR.error 1>> $DEPS_LOG/$LIBWEBSOCKETS_DIR.log
make install 2>> $DEPS_LOG/$LIBWEBSOCKETS_DIR.error 1>> $DEPS_LOG/$LIBWEBSOCKETS_DIR.log
```

```bash
/Users/ccravens/libwebsockets/libwebsockets-2.4.0/lib/http2/http2.c:83:27: error: unused variable 'h2_state_names' [-Werror,-Wunused-const-variable]
static const char * const h2_state_names[] = {
                          ^
1 error generated.
make[2]: *** [CMakeFiles/websockets.dir/lib/http2/http2.c.o] Error 1
make[1]: *** [CMakeFiles/websockets.dir/all] Error 2
make: *** [all] Error 2
/Users/ccravens/libwebsockets/libwebsockets-2.4.0/lib/http2/http2.c:83:27: error: unused variable 'h2_state_names' [-Werror,-Wunused-const-variable]
static const char * const h2_state_names[] = {
                          ^
1 error generated.
make[2]: *** [CMakeFiles/websockets.dir/lib/http2/http2.c.o] Error 1
make[1]: *** [CMakeFiles/websockets.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by lws-team** [comment]
It you use branch v2.4-stable, this along with some other problems are already fixed.

**Comment by ccravens** [comment]
Thanks for your quick response @lws-team. This is using the tagged v2.4.0 release available here: https://github.com/warmcat/libwebsockets/releases/tag/v2.4.0

Is this not the stable version?

**Comment by lws-team** [comment]
No, a tag is a tag.  It's whatever we had on that day in master. The "stable version" that actually gets bugfixes is in this case "v2.4-stable" branch.

What happens is I test stuff from master, according to the list in ./READMEs/release-checklist, it looks good to me, I tag it out.

People start actually using it and testing it and reporting problems on it (like you, they think they should just use whatever is "stable").  Of course, I would like to reverse the order of these events so it gets tested by lots of users and then I make the release.  But nobody responds to my emails about that on the mailing list.

I fix the problems and pile them up into vX.Y-stable branch until it seems to have quietened down again and release a tag v2.4.1, v2.4.2 etc on that.  Since we only put fixes there and no new features, it usually calms down a lot during this process.

So the most stable thing today with a future of getting more stable is v2.4-stable branch.

**Comment by ccravens** [comment]
@lws-team thank you for that explanation!

Ok, just tried the v2.4-stable branch and unfortunately got these errors. I've been building lws for quite sometime, but did a major upgrade tonight to v2.4. I'm wondering is v2.3 a more stable build?

```bash
In file included from /Users/ccravens/libwebsockets/libwebsockets-2.4-stable/plugins/ssh-base/sshd.c:23:
/Users/ccravens/libwebsockets/libwebsockets-2.4-stable/plugins/ssh-base/include/lws-ssh.h:559:1: error: conflicting types for 'timingsafe_bcmp'
timingsafe_bcmp(const void *a, const void *b, uint32_t len);
^
/usr/include/string.h:174:5: note: previous declaration is here
int     timingsafe_bcmp(const void *__b1, const void *__b2, size_t __len);
        ^
/Users/ccravens/libwebsockets/libwebsockets-2.4-stable/plugins/ssh-base/sshd.c:99:1: error: conflicting types for 'timingsafe_bcmp'
timingsafe_bcmp(const void *a, const void *b, uint32_t len)
^
/usr/include/string.h:174:5: note: previous declaration is here
int     timingsafe_bcmp(const void *__b1, const void *__b2, size_t __len);
        ^
2 errors generated.
make[2]: *** [CMakeFiles/protocol_lws_ssh_base.dir/plugins/ssh-base/sshd.c.o] Error 1
make[1]: *** [CMakeFiles/protocol_lws_ssh_base.dir/all] Error 2
make: *** [all] Error 2
In file included from /Users/ccravens/libwebsockets/libwebsockets-2.4-stable/plugins/ssh-base/sshd.c:23:
/Users/ccravens/libwebsockets/libwebsockets-2.4-stable/plugins/ssh-base/include/lws-ssh.h:559:1: error: conflicting types for 'timingsafe_bcmp'
timingsafe_bcmp(const void *a, const void *b, uint32_t len);
^
/usr/include/string.h:174:5: note: previous declaration is here
int     timingsafe_bcmp(const void *__b1, const void *__b2, size_t __len);
        ^
/Users/ccravens/libwebsockets/libwebsockets-2.4-stable/plugins/ssh-base/sshd.c:99:1: error: conflicting types for 'timingsafe_bcmp'
timingsafe_bcmp(const void *a, const void *b, uint32_t len)
^
/usr/include/string.h:174:5: note: previous declaration is here
int     timingsafe_bcmp(const void *__b1, const void *__b2, size_t __len);
        ^
2 errors generated.
make[2]: *** [CMakeFiles/protocol_lws_ssh_base.dir/plugins/ssh-base/sshd.c.o] Error 1
make[1]: *** [CMakeFiles/protocol_lws_ssh_base.dir/all] Error 2
make: *** [all] Error 2
```

**Comment by lws-team** [comment]
What is your platform / compiler?

Neither v2.4.0 or v2.4-stable blow any errors on my box, Travis Linux / Mac or Appveyor Windows.  I guess I will change the name of our timingsafe_bcmp...

**Comment by ccravens** [comment]
My g++ / clang version is:

```bash
Chads-MacBook-Pro-2:libwebsockets ccravens$ g++ --version
Configured with: --prefix=/Applications/Xcode.app/Contents/Developer/usr --with-gxx-include-dir=/usr/include/c++/4.2.1
Apple LLVM version 9.0.0 (clang-900.0.38)
Target: x86_64-apple-darwin16.7.0
Thread model: posix
InstalledDir: /Applications/Xcode.app/Contents/Developer/Toolchains/XcodeDefault.xctoolchain/usr/bin
```

However, I am compiling against a lot of libraries that I've compiled from source rather than using a standard package manager to install. Is this perhaps due to a manually built library that I'm referencing in my build?

**Comment by lws-team** [comment]
I pushed a fix on v2.4-stable... please try it and see if there are any more surprises you left to find with your setup.

**Comment by ccravens** [comment]
@lws-team that seems to have fixed it! I can now successfully build the v2.4-stable branch.

In regards to help testing, is there a wish list of support you need from the community? For better testing, is there an automated build process for lws with unit/functional testing for the library? What are the goals in regards to better testing for the project?

**Comment by lws-team** [comment]
It has CI originally set up by a contributor, every commit gets tested at Travis (including on mac...)

https://travis-ci.org/warmcat/libwebsockets

and Appveyor

https://ci.appveyor.com/project/lws-team/libwebsockets

however lws is very flexible, it has too many combinations of options to test them all in CI, and it works on platforms with no available CI.  Despite turning the warnings up on gcc as far as I can (older gcc does not understand -Wextra, so I can't directly use it and have a selection of older things that enables) and using -Werror, there are still cases like the ones you found that don't get flagged on any of these setups and can only be found by users on their exact toolchain.

I would like to add selftests and run the tests in CI.  But that is tricky when the main test apps want a browser, and it's very timeconsuming to meddle with a remote CI.  So I added a jenkins instance on my server box and may try to go down that road to make my own CI for these things.
